### PR TITLE
[Bug fix] Use shared memory for grad sync when NCCL is not avaliable as PyTorch distributed backend.

### PR DESCRIPTION
**Pull request by classicsong**
## Description
Use shared memory for grad sync when NCCL is not available as PyTorch distributed backend.

Fix small bugs and update unit tests.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Review by nv-dlasalle — APPROVED** [review]

File: python/dgl/optim/pytorch/sparse_optim.py [hunk 1]
```diff
@@ -284,9 +287,11 @@ def _shared_step(self):
                                 # The overall buffer cost will be smaller than three times
                                 # the maximum memory requirement for sharing gradients.
                                 buffer_size = 128 if idx_i.shape[0] < 128 else idx_i.shape[0] * 2
-                                idx_shmem = create_shared_mem_array(idx_shmem_name, \
+                                idx_shmem = create_shared_mem_array(
+                                    '{}_{}'.format(idx_shmem_name, buffer_size), \
                                     (buffer_size,), idx_dtype)
-                                grad_shmem = create_shared_mem_array(grad_shmem_name, \
+                                grad_shmem = create_shared_mem_array(
+                                    '{}_{}'.format(grad_shmem_name, buffer_size), \
```
[the remark this answers is not in the document]
Comment (classicsong) [reply python/dgl/optim/pytorch/sparse_optim.py:294]: When we want to extend the shared memory size, if we use the same name, the old one will be released and will cause some bugs (if the shared memory is freed by the host first, any call to free the same memory in other mirror process will crash the program..

File: src/runtime/cuda/nccl_api.cu [hunk 2]
```diff
@@ -178,7 +178,7 @@ std::pair<IdArray, NDArray> SparsePush(
   // permute the indices and values
   if (num_in > 0) {
     const dim3 block(256);
-    const dim3 grid((num_in+block.x-1)/block.x);
+    const dim3 grid((num_in+block.x-1)/block.x == 0 ? 1 : (num_in+block.x-1)/block.x);
```
Comment (nv-dlasalle) [inline src/runtime/cuda/nccl_api.cu:181]: Now that the `if (num_in > 0) {` was added above, this change is no longer needed.

**Review by nv-dlasalle — APPROVED** [review]
Looks good to me--just the now redundant check for an empty kernel grid, after the other PR was merged.

File: src/runtime/cuda/nccl_api.cu (hunk 2 above)
Comment (classicsong) [reply src/runtime/cuda/nccl_api.cu:181]: I c, will remove it.

**Pull request merged**